Python dependency

datawookie opened the pull request:
Hi!

At present the user is required to manually install the `domino_data` Python package. I have added this to the `.onLoad()` function so that the `domino_data` Python package is installed during the installation of `DominoDataR`.

I have tested these changes in a fresh RStudio workspace on Domino Data Lab.

This MR also contains commits from https://github.com/dominodatalab/r-domino-data/pull/2. Please resolve that first before considering this.

Thanks, Andrew.

daattali commented:
This is excellent, although I would place the code inside a dedicated function (`DominoDataR::setup()` perhaps?) rather than doing so much inside onLoad

I also found out over the weekend that it's possible to do this on installation instead of onLoad.

https://cran.r-project.org/doc/manuals/R-exts.html#Configure-and-cleanup

Here's an example of a configure file that does some setup during installation but only if not inside a CI/CD environment: https://github.com/swarm-lab/Rvision/blob/master/configure

I think this may be a good option. The configure script can call the `setup()` function.

datawookie commented:
@ddl-gabrielhaim and @daattali I have updated the PR.

1. Now uses `configure` script to installed Python libraries at install time (thanks to @daattali for the suggestion).
2. Still uses `.onLoad()` to ensure that the correct Python executable is chose by `{reticulate}`.

@ddl-gabrielhaim could you please review these changes. If you're happy then would be great to merge this in today.

daattali reviewed (commented):
- It looks like your IDE isn't adding a newline character at the end of files. Can you add that, it's good practice to end files with the newline character

- Currently there is no CI in this repo but it's likely that in the future there will be. The python installation doesn't need to happen in CI - we can use `if (nzchar(Sys.getenv('CI'))) { ... }` to make sure the configure script doesn't run in those cases. 

- `scripts/` is not a standard directory for R packages (as far as I'm aware), any non-standard directories should be inside the `inst/` folder

- I still think it'd be a good idea to have the `py_install()` in an exported function rather than a script, so that the user can explicitly call it if needed. Then the configure script just needs to call that function

datawookie commented:
1. I've added `pre-commit`. It, in turn, has fixed all of the files without `\n` at the end.
2. Very happy to set up the CI. Would be good to have some unit tests too, which could also be triggered via CI.
3. Moved the `scripts/` folder under `inst/`.
4. Factored that code out into an exported `py_install()` function.

@daattali is there anything else?

daattali commented:
Cool I've actually never used commit hooks before

The function name `py_install()` should probably be changed to anything else, just so that it's not overwriting reticulate::py_install. If you load both reticulate and this package, then `py_install()` would refer to a different function depending on which one you loaded first. 

Tests would always be welcomed but even without tests the first step is to just run `usethis::use_github_action_check_standard()` in order to make sure the package builds and passes basic smoke tests. (Although I'd do that as a separate PR because it really has nothing to do with the python dependency)

datawookie commented:
Hi @daattali,

1. I have renamed the `py_install()` function. The original name was chosen because this is what you stipulated [here](https://github.com/dominodatalab/DominoDataR/pull/4#pullrequestreview-1190463873).
2. I have added the framework for GitHub actions and created a separate [merge request](https://github.com/dominodatalab/DominoDataR/pull/6). I agree that some tests would be good. But could we just focus on getting this chunk of work merged and then move on to these other things?

Thanks, Andrew.

daattali commented:
I was stuck on this for a few days because I wasn't able to get the package to clearnly install always, and I found the issue and now it's fixed https://github.com/apache/arrow/pull/14744

Thanks for all the help in getting this pacakge to install cleanly, it does seem to work now 🥳 

To answer your two comments:

> I have renamed the py_install() function. The original name was chosen because this is what you stipulated here.

Are you sure you renamed it in this branch? I still see `py_install()` defined and used. Also, what I meant by the earlier comment was that the call to `py_install()` -- referring to `reticulate::py_install()` -- should be done in a function that's exported, so that this setup step can be done manually at any time. For example, if the configure script did not run or succeed for some reason, it'd be good to have a sort of "setup" function that I could run

> could we just focus on getting this chunk of work merged and then move on to these other things

Yep I agree completely, that's exactly what I meant in my comment -- that tests are great but those can come later, even without tests it's still useful to have a GitHub Action that builds the package just to make sure there are no errors!

I also agree with doing that in a separate PR. In a similar vein, I think half the files in this PR should be done separately and removed from this PR to keep the git history clean.

datawookie commented:
@ddl-gabrielhaim I have resolved the merge conflict. Would you mind please merging this today? :pray: Thanks!

daattali commented:
It looks like one of the commits you intended to include here was in a different branch https://github.com/datawookie/DominoDataR/commit/227029f9de130ee80906e3d504710f84158f5aa7 and the precommit hook and lintr shouldn't be bundled in with this PR

Might be best to copy over the changes into a new PR because the commits here are getting messy and it's going to cause a bit of a mess in the history

daattali commented:
I also just noticed that in that misplaced commit, the call to `py_install()` from inside `python-setup.R` wasn't updated to the new function name

datawookie commented:
It has now. See 8464a1284200161cd09f2559b3d4974ce73b16f5.

daattali commented:
🎉  @ddl-gabrielhaim sorry for the huge convo, this should be ready

datawookie commented:
@ddl-gabrielhaim can we please merge this? Thanks.

daattali commented:
@datawookie tests are failing beause the `py_domino_data_install()` is not documented. Can you add documentation to it and run roxygen? I tried doing that myself but because this PR is coming from your git repo and not from a branch on this repository, I'm not sure how to do that

datawookie commented:
Sorted.

daattali commented on R/dominodata.R:
```diff
@@ -0,0 +1,3 @@
+#' @import arrow
+#' @importFrom methods is
+NULL
```
It's good practice to always have a newline at the end of files. I'd suggest configuring your IDE to do that automatically when saving a file

daattali commented on R/python.R:
```diff
@@ -0,0 +1,46 @@
+  # Install the (Python) domino_data package.
+  #
+  if (!reticulate::py_module_available("domino_data")) {
+    result <- try(
+      reticulate::py_install("dominodatalab-data", pip = TRUE, method = "virtualenv")
+    )
+    !is(result, "try-error")
```
```
result <- tryCatch({
  ...
  TRUE
}, error = function(e) {
  FALSE
})
result
```

is a cleaner/more readable version of this.

Also, for future use, `inherits` is preferable to `methods::is()`

daattali commented:
It's strange all these PRs keep being dirtied with merges from other commits, do you know what's causing that? Something in this workflow isn't good and might be worth spending the time to understand how to do cleaner PRs

datawookie commented:
Suggest merging it into branch on target repository and then rebasing against the `main` branch.

daattali commented:
There's still a few small things (the `is()` function is still being imported here, and the configure script should not be running in CI) so I'm going to create a different PR that starts with a clean slate

daattali commented:
closing in favour of #11
